PP-404: Address remaining comments for cgroups hook after checkin

mike0042 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-404](https://pbspro.atlassian.net/browse/PP-404)**

#### Problem description
* Address remaining comments for cgroups hook after checkin

#### Cause / Analysis
* Several changes had to be made to address comments from the original cgroup checkin.

#### Solution description
* Updated the hook and PTL tests accordingly.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [X] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [X] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [X] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bhroam commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -0,0 +1,3657 @@
+                                                  '%s.env')
+        # information for offlining nodes
+        self.offline_file = os.path.join(pbs_home, 'mom_priv', 'hooks',
+               
```
dir() is a python function.  Consider using a different variable name.

bhroam replied:
Same as above, consider using a different variable name.

bhroam replied:
Same as above, consider using a different variable name

bhroam replied:
As above, consider using a different variable name

bhroam replied:
Is is possible for any of these files to not exist?  You've been very careful to try/except all open calls up until now.

bhroam replied:
This comment reads a little weird.  Could you change it to "Check to see if the pid's owner matches the job's owner" or something like that?

bhroam replied:
This log message seems extraneous.

bhroam replied:
Won't 'p' be one of many possible pids in the pids variable?  Why only print one.

bhroam replied:
Re-comment: From what I can tell, 'file' is a reserved word in python. Consider choosing a different variable name.

bhroam replied:
Re-comment: dir() is a python function. Consider using a different variable name.

bhroam replied:
Re-comment: Same as above, consider using a different variable name.

bhroam replied:
Re-comment: Same as above, consider using a different variable name.

bhroam replied:
Same as above, consider using a different variable name.

bhroam replied:
Re-comment: Is is possible for any of these files to not exist? You've been very careful to try/except all open calls up until now.

bhroam replied:
Re-comment: This comment reads a little weird. Could you change it to "Check to see if the pid's owner matches the job's owner" or something like that?

bhroam replied:
Re-comment: This log message seems extraneous.

bhroam replied:
Re-comment: Won't 'p' be one of many possible pids in the pids variable? Why only print one.

mike0042 replied:
It's not uncommon for variables to contain components of other keywords. The keyword "if" is present in most languages and there are all sorts of variables that contain the string. We shouldn't avoid using the variable name "spiffy" because it contains "if".

mike0042 replied:
Same logic applies to avoiding "dir" in variable names. I should be able to use the variable name "dirt" even if it contains "dir".

mike0042 replied:
You lost me on this one.

mike0042 replied:
Lost me again.

mike0042 replied:
Sensing a pattern here.  :-)

mike0042 replied:
Good catch. Not my code, but you are correct.

bhroam replied:
This might have been one of those comments that went to the wrong place.  I don't mind variable names like 'offline_file'.  Using 'file' by itself is a problem.

bhroam replied:
As with the previous one, I'm not talking about a variable like 'dirt'.  Using 'dir' by itself is the issue.  I suspect this was used in a for loop walking directories.

bhroam replied:
I wouldn't try and parse the comments in the wrong place.  I'd only look at the comments which list line numbers that I added as a comment.

bhroam commented:
Weirdness.  Github didn't lose my previous review of the PTL test, it just chose to intermix the comments with the review of the hook I just did.  In any case, I'll try and finish off the review either tomorrow or Monday.

bhroam commented:
Ugh, github lost the location of the last 9 comments.  In conversation they all show up in the same location in the code (lines 1577-1580).  They don't show up in the commit.  It'll be very difficult to determine where they should be located in the code.  Github really fails at large code reviews.

bhroam commented:
OMG, they're all at line 1577 again.  Github is failing.  Time to move this code review outside of github.

mike0042 commented:
Sorry GitHub is causing you grief @bhroam.

bhroam commented:
Here are the last 9 comments per line number:
1265/1766: From what I can tell, 'file' is a reserved word in python. Consider choosing a different variable name.
1767/1845/1946/3209/3217: dir() is a python function. Consider using a different variable name.
1933: Is is possible for any of these files to not exist? You've been very careful to try/except all open calls up until now.
2040: This comment reads a little weird. Could you change it to "Check to see if the pid's owner matches the job's owner" or something like that?
2085: Won't 'p' be one of many possible pids in the pids variable? Why only print one.

bhroam commented:
Here's more comments:
2074: Are you sure you want to reject a job launch because a job has defunct processes?  Why not just ignore them and add the rest to the cgroup?  The defunct processes will get cleaned up at some point.
2079: I don't find this log message too useful.  The following log message pretty much provides the same information.
2129/2132/2139/2142: The job's environment will be large.  You're printing it four times (before and after the add).  If you really want to print the job's environment, I'd only print it once.
2210: Consider combining these three log messages into one.

bhroam commented:
I didn't mean to post those just yet.  Sorry for the slow trickle of comments.
2237/2707: You open a file just to print the contents into a DEBUG3 log message.  Is this really necessary?
2350/2358/2384/2386/2404/2406: No need to use enumerate() if you don't use i
2364/2377/2397(and others): combine these 3 log messages
2374/2394: I'm confused what is going on here.  It looks like you're creating a list.  m.groups(0) is strange because m doesn't exist until you get to one of your subsequent for loops.  You then have 2 for loops to the right of m.groups(0).  Could you explain?
2421-2424(and others): I haven't seen you log the argument values for any other function.  Why here?
2528/2536/2549/2597: Up until now, you've iterated through dictionaries by say for d in dict:.  Here you are saying for d in dict.keys():.  Both are equivalent, but I'd rather see the use be consistent.
2675/2686/2686/2694: Previously you've joined using string.join(list, sep).  Here you use the alternate method of "sep".join(list).  I'd like to see the usage be consistent.
2781: In the previous subsystems, you set resc_used[] in an else case after you see if the value isn't None.  Here you set it further down.  Since this function is repetitious, I'd like to see consistency between the subsystems.
2914: I don't think this comment's content is in the EDD.  I think it might deserve a mention.
2982: Why set attempt?  The for loop on line 2986 will handle that.
2989/2999: attempt has a range of 0-1.  It will always be less than 2 or 3
3039: Why pop if you don't look at its return value?
3050-3051: if statement is duplicated
3140: The comment at the top of the function says it won't sleep.  We sleep for half a second here.
3172: I'm not sure how this log message is useful.  It just prints the internal value of pbs.ND_OFFLINE
3179: Should this be an elif: ?
3247/3359: Consider using a different variable name than 'file'
3267: Is there any purpose to catch any exception just to reraise it?
3372: previously when you wanted to slice from the beginning of an array, you said, arr[:N].  Here you say arr[0:N].  I'd like to see the usage be consistent.
3390: Either uncomment this log message or remove it.  I'd rather not see commented out code in the final hook
3440: Should this comment say 'read in' instead of 'write out'?
3498: I suspect the version you are looking for is the mom version.  Otherwise you could have talked to the server to get the server's version.  If you're going to call out to a separate program to get the version, do a pbs_mom --version (instead of pbs_hostn).
3505: I think this is a left over print statement
3586/3619/3636: Why '(suspended)'?  You are putting a system hold on the job.

I am finally finished with the code review.  Sorry for the mess of code review comments.  Hopefully things won't be too difficult in the future.

mike0042 commented:
Please leave this pull request open. I'll be returning to it soon.

hirenvadalia commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+                                starttime=self.server.ctime)
+        self.remove_vntype()
+        self.wait_and_remove(o.split(':')[1])
+        self.assertTrue(rv)
```
As per latest update in log_match now yo don't need to check return value of log_match so please remove this assertTrue.
This applies to whole file.

mike0042 replied:
What about cleaning up after the call to log_match()? Does log_match() cause the test to complete immediately if the string is not found?

mike0042 replied:
@hirenvadalia Could you please respond to my previous comment?

hirenvadalia replied:
@mike0042 Sorry somehow I missed this reply. Yes log_match() will complete test immediately if given string is not found.

mike0042 replied:
@hirenvadalia If that is the case, then how would one go about cleaning up temporary files after a test case has failed? In this case, we want to test for a specific condition before we cleanup any files.

hirenvadalia replied:
In that you have to use try...expect block and catch exception from log_match and use it to check log existence after your cleanup part.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+        if not self.paths:
+            self.skipTest("No cgroups mounted")
+        self.server.set_op_mode(PTL_CLI)
+        self.server.cleanup_jobs(extend='force')
```
Just making sure we're starting with a clean slate, I guess. If you're certain I should remove it, I will do so.

anamikau replied:
I confirmed that this command is explicitly called by server.revert_to_defaults in PbsTestSuite setUp. I think you don't need it here.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+    iterations = int(sys.argv[1])
+if (len(sys.argv) > 2):
+    chunkSizeMb = int(sys.argv[2])
+if (chunkSizeMb < 1):
```
You are correct. This will be fixed with the next commit.

bhroam commented:
@mike0042 did you make any changes to the hook or did you just move it?  If you made changes and moved the hook it will make the review more difficult.  After you move a file, the entire file shows up as new.  If you did make changes, could you reset the branch before the move and start over?  It shouldn't be hard, just save the file aside, reset the branch, and copy the file back in the old place.  The move of the hook can be the final thing that is done in the review process.  This will allow us to review the hook commit by commit rather than reviewing the entire hook again (which is quite lengthy).

anamikau commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+                                regexp=True,
+                                max_attempts=3,
+                                starttime=self.server.ctime)
+        self.remove_vntype()
```
Correct me if I am wrong but this looks like a cleanup step to me and SetUp is the right place for that before you start any other test.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+                                max_attempts=3,
+                                starttime=self.server.ctime)
+        self.remove_vntype()
+        self.wait_and_remove(o.split(':')[1])
```
We don't have access to the name of the job files in setUp, so we remove them before exiting the test.

anamikau replied:
Ah ok I see.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+        a = {'job_state': 'R'}
+        self.server.expect(JOB, a, jid)
+        self.server.status(JOB, ['Output_Path', 'Error_Path'], jid)
+        (o, e) = (j.attributes['Output_Path'], j.attributes['Error_Path'])
```
No, it doesn't because all of the jobs are submitted with "-j oe". It doesn't hurt anything to set both o and e.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+    def setUp(self):
+        TestFunctional.setUp(self)
+        self.server.expect(SERVER, {'pbs_version': (GE, '14.0')},
+                           max_attempts=3)
```
That assumes you'll always be running the version of PBS that corresponds with the tests, and that's not always the case. On my workstation, I keep a recent version of PTL installed under /opt/ptl and use that to test against the default version of PBS installed. They don't always match, and I don't believe there is any requirement that says they must.

anamikau replied:
Mike, this is a PTL requirement. PTL and PBS should be from same branch. So I agree with Bhroam's comment there.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+        self.hook_name = 'pbs_cgroups'
+        hook_url = 'https://raw.githubusercontent.com/' + \
+            'PBSPro/pbspro/master/src/hooks/cgroups/pbs_cgroups.py'
+        self.load_hook(hook_url)
```
I'll update this to pull the hook from PBS_EXEC/lib/python/altair/pbs_hooks/pbs_cgroups.PY

JonShelley commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -0,0 +1,3657 @@
+                                                  '%s.env')
+        # information for offlining nodes
+        self.offline_file = os.path.join(pbs_home, 'mom_priv', 'hooks',
+               
```
available['cpus'] is a list and not a dictionary. You will need to remove the .keys() or it will fail.

JonShelley replied:
You will need to check to see if mem is available on the socket. Otherwise, if a user request cores that will fit on a single socket and memory of the whole system the job will be killed since it will not be able to access the memory on the other socket due to how cpuset.mems is set to only the socket with the cpus

JonShelley replied:
You will need to check to see if mem is available on the socket. Otherwise, if a user request cores that will fit on a single socket and memory of the whole system the job will be killed since it will not be able to access the memory on the other socket due to how cpuset.mems is set to only the socket with the cpus

JonShelley commented:
On line 2354 we need to remove .keys() since it is a list and not a dictionary
On line 2408 we need to address the memory case since if a job only requests a single socket worth of cores, ngpus or nmics then it will not get the correct values for cpuset.mems which will limit the memory available to the job.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+
+    def random_name(self, length=16):
+        return str().join(random.choice(string.ascii_lowercase)
+                          for _ in range(length))
```
The syntax being used is called list comprehension. It is covered in the Python guide. A performance study is located here: [https://waymoot.org/home/python_string/]

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+            self.skipTest("Freezer cgroup is not mounted")
+        fdir = self.paths['freezer']
+        if not os.path.isdir(fdir):
+            self.skipTest("Freezer cgroup is not found")
```
Moved the checks to the start of the test.

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -406,13 +404,13 @@ def setUp(self):
-        hook_url = 'https://raw.githubusercontent.com/' + \
-            'PBSPro/pbspro/master/src/hooks/cgroups/pbs_cgroups.py'
+        hook_url = 'file:/' + self.server.pbs_conf['PBS_EXEC'] + \
+            '/lib/python/altair/pbs_hooks/pbs_cgroups.PY'
```
Will this fail on Windows due to the slashes?

mike0042 replied:
cgroups are not supported in Windows.

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -451,25 +449,29 @@ def load_hook(self, url):
         hook_file = os.path.join(os.sep, 'tmp', self.random_name())
```
Why use this instead of mkstemp()?

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -451,25 +449,29 @@ def load_hook(self, url):
+                 'content-encoding': 'default',
+                 'input-file': hook_file}
+            rv = self.server.manager(MGR_CMD_IMPORT, HOOK, a, self.hook_name)
+            time.sleep(2)
```
Add a comment explaining the need for the 2 second sleep

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -675,8 +674,7 @@ def test_cgroup_cpuset_and_memory(self):
         except:
-            self.logger.info("Job did not produce expected output")
-            self.assertTrue(False)
+            self.assertTrue(False, "Job did not produce expected output")
```
Don't do assertTrue(False).  Use self.fail(msg) (same for the rest of the test plan)

bhroam replied:
I took a closer look at fail and it doesn't do what we want.  It will fail the test but not include the failure in the failure count at the bottom.  Kind of a useless function if you ask me.  It's really skipping the test, not failing it.

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -899,34 +888,31 @@ def test_cgroup_offline_node(self):
-        if not tasks:
-            self.logger.info("No tasks in cpuset cgroup for job")
-            self.assertTrue(False)
+        self.assertTrue(tasks, "No tasks in cpuset cgroup for job")
```
Since tasks is a list, should you do assertTrue(tasks is not None) ?

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -1064,13 +1049,13 @@ def test_cgroup_reserve_mem(self):
-        vmem = self.server.status(NODE, 'resources_available.vmem')
+        self.server.expect(NODE, {'state': 'free'},
+                           self.momA.shortname, interval=3)
+        vmem = self.server.status(NODEE, 'resources_available.vmem')
```
There's an extra E on the end of NODE

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -451,25 +449,29 @@ def load_hook(self, url):
+            a = {'content-type': 'application/x-python',
+                 'content-encoding': 'default',
+                 'input-file': hook_file}
+            rv = self.server.manager(MGR_CMD_IMPORT, HOOK, a, self.hook_name)
```
Consider using self.server.create_import_hook()

mike0042 commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -0,0 +1,3657 @@
+# "already used" resources
+def job_to_be_ignored(jobid):
+    # Get job substate from small utility that calls printjob
+    pbs_exec = ''
```
We call printjob because it reads job data from the local file rather than contacting the server. Otherwise, the operation scales poorly.

mike0042 commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -0,0 +1,3657 @@
+                                   stderr=subprocess.PIPE)
+        out, err = process.communicate()
+        # Find the job substate
+        substate_re = re.compile(r"substate:\s+(?P<jobstate>\S+)\s+")
```
Not sure why it was done that way. Using group 'jobstate' prevents us from having to split and strip later.

agrawalravi90 commented:
There has been no activity on this for 3 weeks now. @mike0042 do you want to close this for now and re-open it when you get bandwidth to work on this?

mike0042 commented:
Please leave this PR open. It is actively being worked on.

mike0042 commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -0,0 +1,3657 @@
+        numa_nodes = {}
+        for dir in glob.glob(os.path.join(os.sep,
+                             "sys", "devices", "system", "node", "node*")):
+            id = int(dir.split(os.sep)[5][4:])
```
Comment added per request.

mike0042 commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -0,0 +1,3657 @@
+                        threads = []
+                        for xid in cores:
+                            x = cpuinfo['cpu'][xid]
+                            for yid in cores:
```
Comment added per request.

mike0042 commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -0,0 +1,3657 @@
+        # Get the job list from the server
+        jobs = None
+        try:
+            jobs = pbs.server().vnode(self.hostname).jobs
```
This was a very good catch. Querying the server is not scalable. The configure_job method was updated to call gather_jobs_on_node which no longer queries the server.

bhroam commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -2979,85 +2974,73 @@ class CgroupUtils:
-                # cleaned up as well.
-                try:
-                    jobs_list.remove(jobid)
-                    self.cleanup_orphans(jobs_list)
```
Why are we no longer cleaning up orphans?

mike0042 replied:
That's part of the periodic handler now. Code was also added to remove a cgroup if it is already present when the execjob_begin handler is called.

bhroam commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -2979,85 +2974,73 @@ class CgroupUtils:
+            try:
+                hostresc.pop('ncpus')
+            except:
+                pass
```
This would be shorter if you said if 'ncpus' in hostresc: hostresc.pop('ncpus').  2 lines instead of 4.  Plus if you just want to remove 'ncpus' from hostresc, I think using del hostresc['ncpus'] would be more explicit.  pop returns the value which you are ignoring.

mike0042 replied:
Good point. Updated (in next commit).

bhroam commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -2371,6 +2367,7 @@ class CgroupUtils:
             assigned['devices'] = []
             regex = re.compile(".*(mic).*")
             nmics = int(requested['nmics'])
+            # Usa a list comprehension to construct the mics list
```
Usa=Use

mike0042 replied:
Fixed (in next commit).

bhroam commented:
Here are some comments you didn't address.  If you chose not to fix them, please respond so.  I just worry that they were missed.
The line numbers are copied from the original comment:
2210: Consider combining these three log messages into one.
2237/2707: You open a file just to print the contents into a DEBUG3 log message. Is this really necessary?
2350/2358/2384/2386/2404/2406: No need to use enumerate() if you don't use i
    - I saw that you removed one enumerate(), there are still others
2364/2377/2397(and others): combine these 3 log messages
2421-2424(and others): I haven't seen you log the argument values for any other function. Why here?
3172: I'm not sure how this log message is useful. It just prints the internal value of pbs.ND_OFFLINE
3179: Should this be an elif: ?
3498: I suspect the version you are looking for is the mom version. Otherwise you could have talked to the server to get the server's version. If you're going to call out to a separate program to get the version, do a pbs_mom --version (instead of pbs_hostn).

saimanikant commented:
Comments per line number in pbs_cgroups.py : 
(Line numbers are as per the latest commit : 45a1807)
612: The function cgroup.delete() will be called even if there's no possibility of finding any residue of a failed run (like when a job is being run for the first time)
3229: for the above case this func, logs a failed attempt message which is misleading.
754: Can you add a comment about the 3 attempts (at max) if a failure occurs in setting a limit to a subsystem and possible reason for a failure in setting a limit as done in line: 2978  
1167: discover_cpuinfo(): Can you add a comment explaining the structure of 'cpuinfo' returned by this function as done with discover_gpus() or log the same as done in discover_meminfo()
1201: Can you add a comment on why Mapping threads to cores is only possible when value of vendor_id is 'Intel'. Is Hyperthreading only for 'Intel'? What if the value of vendor_id is 'GenuineIntel'?
1558: The function get_vnode_type() is called even if a node is not a vnode. If even "natural" nodes are grouped based on their vntype value, isn't there a possibility of conflict between "run_only_on_hosts" list and "exclude_vntypes" in configuration.

mike0042 commented:
> 2210: Consider combining these three log messages into one.

Two of the messages are now combined and the third is now DEBUG3.

> 2237/2707: You open a file just to print the contents into a DEBUG3 log message. Is this really necessary?

This message was very useful while debugging the initial implementation and is still useful when problems are encountered. These files are in a pseudo-filesystem that is maintained by the kernel with no physical backing store, making them much less expensive to access than disk based storage. It is also useful to capture the actual contents of the file to ensure the contents match what was written.

> 2350/2358/2384/2386/2404/2406: No need to use enumerate() if you don't use i - I saw that you removed one enumerate(), there are still others

There are no longer any instances of enumerate() in the hook.

> 2364/2377/2397(and others): combine these 3 log messages

Fixed.

> 2421-2424(and others): I haven't seen you log the argument values for any other function. Why here?

This is one of the more complex functions to debug and it is helpful to see the value of each argument. Admins must enabled DEBUG3 to see these messages.

> 3172: I'm not sure how this log message is useful. It just prints the internal value of pbs.ND_OFFLINE

You're correct. I removed the message since it really isn't useful.

> 3179: Should this be an elif: ?

It doesn't need to be since the last line of the preceding block will return if the node is offline. Otherwise, we fall through to check whether tmp_state is not None.

> 3498: I suspect the version you are looking for is the mom version. Otherwise you could have talked to the server to get the server's version. If you're going to call out to a separate program to get the version, do a pbs_mom --version (instead of pbs_hostn).

pbs_hostn is the smallest binary we can call to get the local PBS version. It would take more time to invoke pbs_mom.

mike0042 commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -2071,19 +2069,21 @@ class CgroupUtils:
+                # number 1
+                pids = [x for x in pids if x != 1]
         # check pids to make sure that they are owned by the job owner
         if pbs_version[0] >= 13:
```
True, but the cgroups hook is meant to run in a wide variety of configurations. It is likely an administrator will replace their existing cgroups hook with a newer one, even if they are running a 12.x version.

bhroam replied:
The cgroups hook that ships with 17 will work in a certain way.  The cgroups hook that ships in a later release will work in a new way.  The admin should use the hook provided with the release they have installed like with the rest of PBS works.  You don't suddenly replace qsub with a newer version.  Why should it work the same with the cgroups hook. 
 We shouldn't try and write one hook that runs everywhere.  It gets very ugly very fast.

mike0042 commented:
Addressing comments from @saimanikant. Changes will be present in next commit.

> 612: The function cgroup.delete() will be called even if there's no possibility of finding any residue of a failed run (like when a job is being run for the first time)

That's correct, and it is OK because the delete method calls the remove_cgroup method and simply returns if the cgroup is not present.

> 3229: for the above case this func, logs a failed attempt message which is misleading.

Good point. Updated the method logic to log the message only when the removal actually fails.

> 754: Can you add a comment about the 3 attempts (at max) if a failure occurs in setting a limit to a subsystem and possible reason for a failure in setting a limit as done in line: 2978

Comment added.

> 1167: discover_cpuinfo(): Can you add a comment explaining the structure of 'cpuinfo' returned by this function as done with discover_gpus() or log the same as done in discover_meminfo()

Added comments and added log message to discover_cpuinfo.

> 1201: Can you add a comment on why Mapping threads to cores is only possible when value of vendor_id is 'Intel'. Is Hyperthreading only for 'Intel'? What if the value of vendor_id is 'GenuineIntel'?

Yes, hyperthreading is only for processors that identify as "Intel" or "GenuineIntel". The string "Intel" is in "GenuineIntel", so either string will trigger the condition.

> 1558: The function get_vnode_type() is called even if a node is not a vnode. If even "natural" nodes are grouped based on their vntype value, isn't there a possibility of conflict between "run_only_on_hosts" list and "exclude_vntypes" in configuration.

There is a precedence here, so there is no conflict. If a host is listed in "run_only_on_hosts" it overrides the exclude lists for vnode type and hosts.

mike0042 commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -2071,19 +2069,21 @@ class CgroupUtils:
+                # number 1
+                pids = [x for x in pids if x != 1]
         # check pids to make sure that they are owned by the job owner
         if pbs_version[0] >= 13:
```
It does get ugly fast, but we have early adopters that want the latest version of the hook regardless of PBS version they are running. While qsub is a compiled command, hooks are meant to be modifiable. In general, hooks should work across a wide range of PBS versions unless development has deliberately broken compatibility.

bhroam replied:
PBS hooks are considered core PBS functionality.  I still think this means if a new version of the hook comes out, it will depend on the version that it was shipped with.  We won't go out of our way to break it, but breakage can happen.  An example is that you usually can use a newer version of the scheduler with an older server.  We don't go out of our way to break that communication, but some times the communication does break.  To fix this, PBS has to be upgraded as a whole.  We don't have anything built into that communication pathway to allow for this.  I don't think we should do it in hooks either.  Not to mention that this is an OSS checkin and the earliest version of OSS is 14.1.  You're checking for >= 13.

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1083 @@
+        self.hook_name = 'pbs_cgroups'
+        hook_url = 'https://raw.githubusercontent.com/' + \
+            'PBSPro/pbspro/master/src/hooks/cgroups/pbs_cgroups.py'
+        self.load_hook(hook_url)
```
Why pull the hook at all?  Just enabled it.  It's a PBS hook, and should be installed when PBS is installed.

bhroam commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -3018,10 +3021,8 @@ class CgroupUtils:
             # Remove the ncpus key if it exists. Ignore any KeyError.
-            try:
+            if 'ncpus' in hostresc:
                 hostresc.pop('ncpus')
```
Change this to del hostresc['ncpus']

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -392,7 +392,8 @@ def setUp(self):
         a = {'log_events': '2047'}
-        self.server.manager(MGR_CMD_SET, SERVER, a, expect=True)
+        rv = self.server.manager(MGR_CMD_SET, SERVER, a, expect=True)
+        self.assertEqual(rv, 0)
```
No need to check the return code of manager().  It will raise an exception if it fails (same fore the rest of the tests).

bhroam commented:
One point to ponder.  DEBUG3 messages are used by the mom for debugging purposes too.  The vast number of DEBUG3 messages that are in this hook will make it very difficult to debug the mom itself.

mike0042 commented:
Would you suggest DEBUG4 @bhroam?

anamikau commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        self.server.expect(NODE, {'state': 'free'},
+                           self.momA.shortname, interval=3)
+        vmem = self.server.status(NODE, 'resources_available.vmem')
+        self.logger.info("%s" % str(vmem))
```
modify the message to say that this is vmem value

mike0042 replied:
Done.

anamikau commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        conf = {'enabled': 'False', 'freq': 30}
+        rv = self.server.manager(MGR_CMD_SET, HOOK, conf, self.hook_name)
+        self.assertEqual(rv, 0)
+        TestFunctional.tearDown(self)
```
This line should be the first line in tearDown function

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        # Increase the log level
+        a = {'log_events': '2047'}
+        rv = self.server.manager(MGR_CMD_SET, SERVER, a, expect=True)
+        self.assertEqual(rv, 0)
```
These were taken care of based on one of @bhroam's comments. Changes will be present in the next commit.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        self.hook_name = 'pbs_cgroups'
+        hook_url = 'file:' + self.server.pbs_conf['PBS_EXEC'] + \
+            '/lib/python/altair/pbs_hooks/pbs_cgroups.PY'
+        self.load_hook(hook_url)
```
No, PTL does a pretty good job making sure site hooks are deleted. This is now updated to pull the hook from its installed location.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+            self.assertTrue(rv)
+            # Add the configuration
+            self.load_config(self.cfg0)
+            # Now load the hook
```
It appears to be redundant, so I removed it.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+             'input-file': cfg_file}
+        rv = self.server.manager(MGR_CMD_IMPORT, HOOK, a, self.hook_name)
+        os.remove(cfg_file)
+        time.sleep(2)
```
I didn't understand it either, so I removed them.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+                pass
+            break
+
+    def xtest_cgroup_vntype_excluded(self):
```
Good catch. Because those tests are currently being skipped.

bhroam replied:
How about skipping them with self.skipTest()?

anamikau replied:
They are valid test so we don't want to skip them. I think we need to analyze and if there is a PTL/PBS bug then only skip them

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        that have an exclude vntype file set
+
+        This test is disabled because it times out while
+        running with other tests, but runs fine by itself.
```
I'm not sure why either. One of the mysteries of PTL?

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+                               starttime=self.server.ctime)
+        except:
+            rv = False
+        self.wait_and_remove(o.split(':')[1])
```
Because it's no longer needed and all tests should clean up after themselves.

anamikau replied:
you are right about cleaning up. What if the test fails prior to this line. Who will cleanup then?

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        except:
+            rv = False
+        self.wait_and_remove(o.split(':')[1])
+        self.assertTrue(rv)
```
Yes, no need to set rv here. The test will fail if any log_match fails.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        """
+        Test to verify that the job cgroup is created correctly
+        Check to see that cpuset.cpus=0, cpuset.mems=0 and that
+        memory.limit_in_bytes = 104857600
```
Comment updated.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        self.server.status(JOB, [ATTR_o, ATTR_e], jid)
+        (o, e) = (j.attributes[ATTR_o], j.attributes[ATTR_e])
+        tmp_out = ''
+        try:
```
OK

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+            with open(tmp_file, 'r') as fd:
+                tmp_out = fd.read().splitlines()
+        except:
+            self.assertTrue(False, "Job did not produce expected output")
```
Looks accurate to me. What seems inaccurate?

anamikau replied:
the inaccuracy is above error message says "Job did not produce expected output" however you are checking the output below after that. The accurate message would be "job output file not found or do not have enough permissions to read"

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        a = {'job_state': 'R'}
+        self.server.expect(JOB, a, jid)
+        self.server.status(JOB, [ATTR_o, ATTR_e], jid)
+        (o, e) = (j.attributes[ATTR_o], j.attributes[ATTR_e])
```
Removed them all.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        (o, e) = (j.attributes[ATTR_o], j.attributes[ATTR_e])
+        rv = True
+        try:
+            self.mom.log_match(".*%s;Cgroup mem\w+ limit exceeded.*" % jid,
```
Updated for multiple tests.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        try:
+            fdir = os.path.join(fdir, 'pbspro')
+            if not os.path.isdir(fdir):
+                os.mkdir(fdir)
```
This isn't a distributed test. Why use the distributed commands?

anamikau replied:
@hirenvadalia has commented on other PR's to use PTL dshutils command for such OS operations. I am sure there is a good reason for that. I will let @hirenvadalia clarify.

hirenvadalia replied:
@mike0042 Are you sure that whatever job you are submitting in this test will run only on host on which you have invoked pbs_benchpress command? If yes then, you don't need to use methods from DshUtils... else you need to use along with passing hostname as where your job is running...

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -0,0 +1,1120 @@
+        time.sleep(1)
+        try:
+            os.rmdir(fdir)
+            os.rmdir(os.path.dirname(fdir))
```
This isn't a distributed test. Why use the distributed commands?

mike0042 commented:
@bhroam @anamikau @saimanikant @JonShelley 
One more commit has just been pushed, and I expect we're close to completing this PR. Please provide any additional feedback you may have.

bhroam commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -2418,11 +2430,11 @@ class CgroupUtils:
+        pbs.logmsg(pbs.EVENT_DEBUG4, "Requested: %s" % (requested))
+        pbs.logmsg(pbs.EVENT_DEBUG4, "Available: %s" % (available))
+        pbs.logmsg(pbs.EVENT_DEBUG4, "Numa Nodes: %s" % (node.numa_nodes))
+        pbs.logmsg(pbs.EVENT_DEBUG4, "Devices: %s" % (node.devices))
```
Why print 4 lines to the log when one would suffice?

mike0042 replied:
We still want devices to be on its own line because it could be quite long.

bhroam commented on src/hooks/cgroups/pbs_cgroups.PY:
```diff
@@ -2563,10 +2575,10 @@ class CgroupUtils:
+                pbs.logmsg(pbs.EVENT_DEBUG4, "cpus: %s" % cpus)
+                pbs.logmsg(pbs.EVENT_DEBUG4, "sockets: %s" % sockets)
+                pbs.logmsg(pbs.EVENT_DEBUG4, "devices: %s" % devices)
+                pbs.logmsg(pbs.EVENT_DEBUG4, "memory limit: %sB" % memory)
```
Why print 4 lines to the log when one would suffice?

mike0042 replied:
We still want devices to be on its own line because it could be quite long.

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -392,8 +391,7 @@ def setUp(self):
         Job.dflt_attributes[ATTR_k] = 'oe'
         # Increase the log level
         a = {'log_events': '2047'}
```
Now that we changed the debug messages to debug4, does this need to change?

mike0042 replied:
PBSEVENT_DEBUG4 is defined as 0x0800 (which is 2048), so yes, I'll update that.

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -450,19 +449,18 @@ def random_name(self, length=16):
-        self.assertTrue(script != '')
-        hook_file = os.path.join(os.sep, 'tmp', self.random_name())
-        with open(hook_file, "w") as fd:
+        self.assertNotEqual(script, '')
```
Why don't you put this except in the except block?  It's when script gets set to ''

mike0042 replied:
Good point.

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -528,13 +517,11 @@ def wait_and_remove(self, filename):
             except:
-                pass
-            break
+                time.sleep(1)
```
Do you want to complain if you did this 10 times and the file is still not deleted?

mike0042 replied:
IMHO, no sense in complaining if there's nothing that can be done.

bhroam replied:
Will your test fail?  If so, it might be nice to have some indication of why the test failed.

bhroam commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -586,19 +568,14 @@ def xtest_cgroup_host_excluded(self):
+                           "\['%s'\].*" % self.mom.shortname,
+                           regexp=True,
+                           max_attempts=5,
+                           starttime=self.server.ctime)
```
log_match() does a substring match.  You can remove the .* at the front and end and remove regexp=True.  I see other log_match() calls like this.  Fix those too.

mike0042 replied:
I fixed those that were not making use of regular expressions in some other way.

bhroam replied:
It is not clear to me if you are going to fix this.  This log_match() does not require regex.  All you are saying is match any number of characters before and after your string.  log_match() already does substring matching, so you don't need regex to do that.

mike0042 replied:
I removed the "regex=True" from a few of them that didn't need it. New commit coming soon.

anamikau commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -471,21 +469,14 @@ def load_hook(self, url):
-                                                overwrite=True)
-            self.assertTrue(rv)
+            self.server.create_import_hook(self.hook_name, a, script,
+                                           overwrite=True)
```
I am not convinced that we need a new function just for create and importing hook. we will avoid reading the hook file into a new file and rather move line 466-472 in setUp. And again why we need try-catch for everything rather than taking the output into a variable and using assertTrue.

mike0042 replied:
We need this in a try block so that the temp file gets removed within the finally section. That will run regardless of whether the load is successful.

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -6362,7 +6362,7 @@ def manager(self, cmd, obj_type, attrib=None, id=None, extend=None,
                 if cmd == MGR_CMD_LIST:
                     bsl = self.utils.convert_to_dictlist(ret['out'], attrib,
-                                                         mergelines=False)
+                                                         mergelines=True)
```
@anamikau Even I don't know why it was False in first place... I just quick checked first version of PTL in P4 and there also it is False so I believe previously we never got as any output of qmgr list command which has one line wrapped to multiple line...

mike0042 replied:
I can't explain why it was false to begin with. It was like that when it was handed off. The event list is split into multiple lines when it gets long. I have not filed a bug because I don't think PTL was doing anything wrong.

mike0042 commented on test/tests/functional/pbs_cgroups_hook.py:
```diff
@@ -404,23 +402,24 @@ def setUp(self):
-                                 runas=ROOT_USER, logerr=False)
+        self.server.manager(MGR_CMD_CREATE, RSC, attr, id='nmics',
+                            runas=ROOT_USER, logerr=False)
         self.assertEqual(rv, 0)
```
Fixed a couple of these.

bhroam commented:
I have no more comments on the current commits, except that we never finished our conversation about checking the pbs version inside the hook.  I am still against it.  The hook is shipped with PBS, so the version of the hook that is shipped with PBS should be the one used with that version of PBS.  When a new version of PBS is installed, the new hook that came with the version should be used.  Even if you don't buy into that argument, you're checking for pbs version < 13.  Versions less than 13 are not supported by either the OSS project or by Altair.  There is no reason the hook should need to check for it.

mike0042 commented:
@bhroam we can certainly discuss removing the PBS version check, but at one point it was being used on older versions of PBS. It doesn't harm anything to leave it there if you are running a recent version. Shall I file a ticket, or is this something you feel should not be introduced in the first place?

bhroam commented:
In the case of checking pbs version < 13, it's basically dead code.  I see no reason to introduce it in the first place.

saimanikant commented:
Changes looks good to me. I sign off.

mike0042 commented:
@bhroam: @JonShelley indicated that the check may be removed, so I have gone ahead and done so. Please review this latest minor change and provide your signoff. I'll also need signoff from @saimanikant. I'll have a Jenkins build ready for @anamikau shortly.

arungrover commented:
hey @mike0042  I see you have 2 sign-off now. In the interest of time please squash and rebase

mike0042 commented:
@arungrover @bhroam 
Squashed and rebased. CI checks are in progress.

mike0042 commented:
@arungrover @bhroam
CI checks have passed.

bhroam commented:
I merged the PR

mike0042 commented:
Thank you @bhroam.
